server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4780,14 +4795,57 @@ private String copySegmentFiles(
                     storeDirectory.deleteFile(file);
                 }
             }
+            // Compute the number of files to download
+            // Required so that we don't publish metrics if there are no files to copy over
             for (String file : uploadedSegments.keySet()) {
                 long checksum = Long.parseLong(uploadedSegments.get(file).getChecksum());
                 if (overrideLocal || localDirectoryContains(storeDirectory, file, checksum) == false) {
-                    storeDirectory.copyFrom(sourceRemoteDirectory, file, file, IOContext.DEFAULT);
-                    downloadedSegments.add(file);
+                    segmentsToDownload.add(file);
                 } else {
                     skippedSegments.add(file);
                 }
+            }
+            // Upload stats before starting Segment downloads
```

Comment:
This stats logic need to be abstracted out from this method. You can have methods like `beforeDownload`, `afterDownload` and have `copyFrom` which encapsulates file level copy along with stats.

Reply:
Ack. Working on this.